src/adm.rs

```diff
@@ -1,39 +1,269 @@
+    ) -> HandlerResult<()> {
+        check_url!(&tile.click_url, "Click", filter.click_hosts, tags)
+    }
+
+    /// Check the impression URL to see if it's valid.
+    ///
+    /// This extends `filter_and_process`
+    fn check_impression(
+        &self,
+        filter: &AdmAdvertiserFilterSettings,
+        tile: &mut AdmTile,
+        tags: &mut Tags,
+    ) -> HandlerResult<()> {
+        check_url!(
+            &tile.impression_url,
+            "Impression",
+            filter.impression_hosts,
+            tags
+        )
+    }
+
+    /// Filter and process tiles from ADM:
+    ///
+    /// - Returns None for tiles that shouldn't be shown to the client
+    /// - Modifies tiles for output to the client (adding additional fields, etc.)
+    pub fn filter_and_process(&self, tile: AdmTile, tags: &mut Tags) -> Option<AdmTile> {
+        // Use strict matching for now, eventually, we may want to use backwards expanding domain
+        // searches, (.e.g "xyz.example.com" would match "example.com")
+        let mut result = tile.clone();
+        match self.filter_set.get(&tile.name.to_lowercase()) {
+            Some(filter) => {
+                // Apply any additional tile filtering here.
+                let none = AdmAdvertiserFilterSettings::default();
+                let default = self
+                    .filter_set
+                    .get(&DEFAULT.to_lowercase())
+                    .unwrap_or(&none);
+                let adv_filter = if filter.advertiser_hosts.is_empty() {
+                    default
+                } else {
+                    filter
+                };
+                let impression_filter = if filter.impression_hosts.is_empty() {
+                    default
+                } else {
+                    filter
+                };
+                let click_filter = if filter.click_hosts.is_empty() {
+                    default
+                } else {
+                    filter
+                };
+                match self.check_advertiser(adv_filter, &mut result, tags) {
+                    Ok(_) => {}
+                    Err(e) => {
+                        self.report(&e, tags);
+                        return None;
+                    }
+                }
+                match self.check_click(click_filter, &mut result, tags) {
+                    Ok(_) => {}
+                    Err(e) => {
+                        self.report(&e, tags);
+                        return None;
+                    }
+                }
+                match self.check_impression(impression_filter, &mut result, tags) {
+                    Ok(_) => {}
+                    Err(e) => {
+                        self.report(&e, tags);
+                        return None;
+                    }
+                }
+                result.position = if filter.position.is_none() {
+                    default.position
+                } else {
+                    filter.position
+                };
```

Comment:
Could this use `unwrap_or`?

```rust
result.position = filter.position.unwrap_or(default.position);
```

Reply:
Actually, gonna loop back on this one, because It's a bit subtle and I wrote it at 10:00pm, which is never a good thing.

I can't really use "unwrap" here because I don't actually want the unwrapped content. I want the fact that this is an `Option<u8>`. Doing the check the way I specified just lets me check if I should use the `default` value if `filter` isn't set. Both could be `None`.

Reply:
You want `result.position = filter.position.or(default.position)`